packages/ra-ui-materialui/src/field/ArrayField.js

```diff
@@ -0,0 +1,140 @@
+import { Component, cloneElement } from 'react';
+import PropTypes from 'prop-types';
+import get from 'lodash/get';
+import pure from 'recompose/pure';
+
+const initialState = {
+    data: {},
+    ids: [],
+};
+
+/**
+ * Display a collection
+ * 
+ * Ideal for embedded arrays of objects, e.g.
+ * {
+ *   id: 123
+ *   tags: [
+ *     { name: 'foo' },
+ *     { name: 'bar' }
+ *   ]
+ * }
+ * 
+ * The child must be an iterator component
+ * (like <Datagrid> or <SingleFieldList>).
+ * 
+ * @example Display all the backlinks of the current post as a <Datagrid>
+ * // post = {
+ * //   id: 123
+ * //   backlinks: [
+ * //       {
+ * //           date: '2012-08-10T00:00:00.000Z',
+ * //           url: 'http://example.com/foo/bar.html',
+ * //       },
+ * //       {
+ * //           date: '2012-08-14T00:00:00.000Z',
+ * //           url: 'https://blog.johndoe.com/2012/08/12/foobar.html',
+ * //       }
+ * //    ]
+ * // }
+ *     <ArrayField source="backlinks">
+ *         <Datagrid>
+ *             <DateField source="date" />
+ *             <UrlField source="url" />
+ *         </Datagrid>
+ *     </ArrayField>
+ *
+ * @example Display all the tags of the current post as <Chip> components
+ * // post = {
+ * //   id: 123
+ * //   tags: [
+ * //     { name: 'foo' },
+ * //     { name: 'bar' }
+ * //   ]
+ * // }
+ *     <ArrayField source="tags">
+ *         <SingleFieldList>
+ *             <ChipField source="name" />
+ *         </SingleFieldList>
+ *     </ArrayField>
+ *
+ * If you need to render a collection in a custom way, it's often simpler
+ * to write your own component:
+ * 
+ * @example
+ *     const TagsField = ({ record }) => (
+ *          <ul>
+ *              {record.tags.map(item => (
+ *                  <li key={item.name}>{item.name}</li>
+ *              ))}
+ *          </ul>
+ *     )
+ *     TagsField.defaultProps = { addLabel: true };
+ */
+export class ArrayField extends Component {
```

Comment:
As this component does not include any UI, shouldn't it be in `ra-core` ?

Reply:
I though about that, too. It's just that we don't have the folder structure in `ra-core` to accommodate plain fields (not just controllers). I was being lazy, I'll do it.

Reply:
But the `ArrayField.spec.js` file does use many components from `ra-ui-material-ui`. In fact, it tests the whole integration with an iterator. So it's not easy to move the component and its tests to `ra-core`.

We need to think more about what goes into core and what goes into ui. But in a future PR.

Reply:
ok